src/lightning/pytorch/trainer/trainer.py

```diff
@@ -965,33 +949,16 @@ def _teardown(self) -> None:
         self._signal_connector.teardown()
 
     def _run_stage(self) -> Optional[Union[_PREDICT_OUTPUT, _EVALUATE_OUTPUT]]:
-        # wait for all to join if on distributed
-        self.strategy.barrier("run-stage")
```

Comment:
`checkpoint_connector.resume_end` already does a barrier at its end

Reply:
I highly recommend to leave this here. Here is the best place to do it, to guarantee that a stage always starts synchronous with all processes. We should not rely on the checkpoint connector to do it at the end, it would be error prone to rely on a completely independent component to do it, just because it happens to be called before run stage.

Reply:
> Barrier'ing after loading also seems sensible.

Absolutely, both barriers must stay.

This is to be safe yes, and for correctness. The amount of time spent finding and learnings from the past to get these barriers in the right place is too precious to risk someone refactoring "checkpointing" and causing the disappearance of a barrier, which would then lead to mysterious issues over several releases until someone finds the issue.

> Barriers go against performance

Not here, this runs once at the beginning. And compared to what comes before (setup, checkpoint loading etc.), this barrier will not be noticeable.